File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode2210CountHillsAndValleysInAnArray.cpp
#include "LeetCode2210CountHillsAndValleysInAnArray.h"

#include <iostream>
using namespace std;

int LeetCode2210CountHillsAndValleysInAnArray::countHillValley(vector<int>& nums)
{
    long slope = 0;
    long newSlope = 0;
    int result{};
    int previousNum = nums[0];
    for (int i = 1, e = nums.size(); i < e; i++)
    {
        if (nums[i] == previousNum)
            continue;
        newSlope = nums[i] - previousNum;
        previousNum=nums[i];
        if (newSlope * slope < 0)
            result++;
        slope = newSlope;
    }
    return result;
}

void LeetCode2210CountHillsAndValleysInAnArray::Run()
{
    vector<int> nums= {2, 4, 1, 1, 6, 5};
    cout<<countHillValley(nums)<<endl;
}

File: LeetCodeProblemsGo/LeetCode2210CountHillsAndValleysInAnArray/LeetCode2210CountHillsAndValleysInAnArray.go
package LeetCode2210CountHillsAndValleysInAnArray

func countHillValley(nums []int) int {
	var lastSlope int64 = 0
	var slope int64 = 0
	previousAmount := nums[0]
	length := len(nums)
	result := 0
	for i := 1; i < length; i++ {
		if nums[i] == previousAmount {
			continue
		}
		slope = int64(nums[i] - previousAmount)
		previousAmount = nums[i]
		if lastSlope*slope < 0 {
			result++
		}
		lastSlope = slope
	}
	return result

}
